From 78ce48f6665db4d84fca89bb42bf73233772786d Mon Sep 17 00:00:00 2001
From: Wolfgang Haupt <haupt.wolfgang@gmail.com>
Date: Thu, 11 Jul 2019 18:24:08 +0200
Subject: [PATCH] VideoPlayer: Properly propagate playback errors

Since quite some time, we do not show the playback error dialog
box anymore, when we fail to create a demuxer.

As the m_bAbortRequest-Flag has precedence over the m_error flag,
when we choose between OnPlaybackStopped, OnPlaybackError or
OnPlaybackEnded, we ended up never reporting playback errors.

To fix this issue, we get rid of the double function that
m_bAbortRequest serves, by introducing a m_bCloseRequest-Flag.
This flag is set in CVideoPlayer::CloseFile and used
instead of the m_bAbortRequest-Flag, when we chose callbacks.
---
 xbmc/cores/VideoPlayer/VideoPlayer.cpp | 7 +++++--
 xbmc/cores/VideoPlayer/VideoPlayer.h   | 1 +
 2 files changed, 6 insertions(+), 2 deletions(-)

diff --git a/xbmc/cores/VideoPlayer/VideoPlayer.cpp b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
index eb380e68d45f..65d813a2c41b 100644
--- a/xbmc/cores/VideoPlayer/VideoPlayer.cpp
+++ b/xbmc/cores/VideoPlayer/VideoPlayer.cpp
@@ -636,6 +636,7 @@ CVideoPlayer::CVideoPlayer(IPlayerCallback& callback)
 
   m_displayLost = false;
   m_error = false;
+  m_bCloseRequest = false;
   CServiceBroker::GetWinSystem()->Register(this);
 }
 
@@ -675,6 +676,7 @@ bool CVideoPlayer::OpenFile(const CFileItem& file, const CPlayerOptions &options
   m_processInfo->SetPlayTimes(0,0,0,0);
   m_bAbortRequest = false;
   m_error = false;
+  m_bCloseRequest = false;
   m_renderManager.PreInit();
 
   Create();
@@ -691,6 +693,7 @@ bool CVideoPlayer::CloseFile(bool reopen)
 
   // set the abort request so that other threads can finish up
   m_bAbortRequest = true;
+  m_bCloseRequest = true;
 
   // tell demuxer to abort
   if(m_pDemuxer)
@@ -2443,9 +2446,9 @@ void CVideoPlayer::OnExit()
   m_bStop = true;
 
   bool error = m_error;
-  bool abort = m_bAbortRequest;
+  bool close = m_bCloseRequest;
   m_outboundEvents->Submit([=]() {
-    if (abort)
+    if (close)
       cb->OnPlayBackStopped();
     else if (error)
       cb->OnPlayBackError();
diff --git a/xbmc/cores/VideoPlayer/VideoPlayer.h b/xbmc/cores/VideoPlayer/VideoPlayer.h
index b26ccc6873ab..8661a172c137 100644
--- a/xbmc/cores/VideoPlayer/VideoPlayer.h
+++ b/xbmc/cores/VideoPlayer/VideoPlayer.h
@@ -443,6 +443,7 @@ class CVideoPlayer : public IPlayer, public CThread, public IVideoPlayer,
   CPlayerOptions m_playerOptions;
   bool m_bAbortRequest;
   bool m_error;
+  bool m_bCloseRequest;
 
   ECacheState  m_caching;
   XbmcThreads::EndTime m_cachingTimer;
